nftables formula

tacerus opened the pull request:
New formula for managing nftables. Work in progress.

tacerus committed:
Initialize nftables formula

First concept idea.

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus commented on nftables-formula/pillar.example:
Hi @cboltz, what do you think of this first idea of a pillar structure? I'm not a big fan of the "priority" keys under the chain entries, but Salt does not support loading the pillar as an ordered dictionary.

cboltz replied:
First of all, note that `priority: 01` leads to the filename `1_variables.conf` so the leading zero gets lost. That might cause some trouble later.

Also, I'm not sure if the priority for the main sections (variables, tables) is needed because variable declarations always have to come first. However, we need something for ordering the actual rules, because there order is relevant. (And maybe we even need ordering inside the variable declaration if a variable includes another one.)

I understand why you don't like the priority keys, but I'm afraid we need them (or something similar).

Just as an idea, and I'm not sure if it would be better: What about

```
              rules:
                201:
                  'ct state established, related':
                    action: accept
                202:
                  'ct state invalid':
                    action: drop
```

or even more boring:

```
              rules:
                201: 'ct state established, related accept'
                202: 'ct state invalid drop'
```

I know the boring variant comes with a few (IMHO: theoretical) restrictions, but the better readability might be worth it.

As a side note: Your rendered example lacks the action part ("accept"/"drop") for the `ct ...` rules.

I also wonder if splitting between `rules:` and `meta:` makes sense, or if you should group both in the same section. In the end, they get merged anyway - and understanding the pillar might be easier if they aren't split into two areas.

tacerus replied:
Thank you for the input. Using numbers as keys seems to be similar to what Ansible is doing (https://github.com/ipr-cnrs/nftables/blob/master/defaults/main.yml#L325).

Another idea would be use lists, it would still require ordering of the lists but only on a table and chain level.

One benefit I see of splitting the rule lines into different YAML keys is easier parsing by other tools. But that might not be needed.

I also found nftables supporting a native JSON format, but it might require to write Salt modules to make use of it using Python. It seems to utilize lists for ordering as well but the YAML equivalent might not be as user friendly in a Salt pillar.

> Your rendered example lacks the action part 

Thanks, culprit is a stray whitespace.

tacerus replied:
It seems the pillar in modern versions of Salt is actually stored in an `OrderedDict` object. Though the list based approach would still be easier.

tacerus committed:
Use list based nftables rules

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus committed:
Last mako version

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus committed:
Refactor

- use Jinja instead of Mako template (additional Python logic is no longer needed)
- split into active and passive configurations for includes which shouldn't happen on the top-level
- add file header
- support sets

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus committed:
Move logic to macros, improve includes

- Allows the same loops to be used both in the top-level as well as in lower contexts.
- Expand includes to their file path if they match a passive configuration.

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus committed:
Drop mako template

No longer used in favor of the Jinja one.

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus committed:
Support vmaps

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus committed:
Support maps

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus committed:
Support multiple includes

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus committed:
Make table type optional

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus committed:
Support multiple chain jumps

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus committed:
Support counter and log in chains

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

darix commented on nftables-formula/pillar.example:
```diff
@@ -0,0 +1,56 @@
+            - ct state established, related accept
+    sets:
+      mode: passive
+      sets:
```
there needs to be a layer for inet/ip/ip6 here.

tacerus replied:
Can you elaborate, do you mean some abstraction which will generate sets for the different families? Currently I just build `set`s the same way as natively in nftables (see the macros file https://github.com/openSUSE/salt-formulas/pull/61/files/730c44859d4d76bb02a4c3b17dc371161695382f), which intends specifying the family as part of the set `type`.

darix replied:
how do you decide if a set should go into 

```
table ip6 nat {
}
```

```
table ip nat {
}
```

which is might be needed if you want to use sets in nat rules . those can not live in `table inet ...`

tacerus replied:
That depends on where you include the file you define the set in. In the example pillar, a configuration file `sets.conf` is placed in `/etc/nftables.d/passive/`, which is then included in the `inet` table via line 34.

cboltz commented:
using https://github.com/darix/madeyougoaway might be an alternative to this draft PR.
